### ISSUE TITLE
Release 2022-03 follow-up

### ISSUE THREAD

**Issue by stonebig** [winpython/winpython#1085]
release: october 2022

Wanted:
- Infrastructure:
   - [x] cpython-3.10.6
   - [x] cpython-3.11.0 (10% to 25% speed-up vs 3.10 on interactive Python applications)
   - [x] github as a full secondary download site
   - no more MKL with pypi, for now, because:
      - loss of [cgohlke site](https://www.lfd.uci.edu/~gohlke/pythonlibs/)
      - no failover found.

Focus:
- python-3.11.0 stack debuging: 
   - [x] patch patsy
-  infrastructure work:
   - [x] re-implement pipdeptree into wppm, proof-of-concept in docs [as a notebook ](https://github.com/winpython/winpython_afterdoc/blob/master/docs/pipdeptree_replacement.ipynb)
 - more arrow/rust/wasm, less PyPy/mkl

Postpone:
   - [ ] move away from setup.py https://setuptools.pypa.io/en/latest/userguide/quickstart.html#transitioning-from-setup-py-to-setup-cfg (requires change in build system)
   - PyPy-3.9 (loss of Christoph Gohlke wheels)

 
Hopes/Wishes for 2022/2023:
- Python speed-up movement: 
   - per-thread GIL: [Shannon plan for 3.12](https://github.com/faster-cpython/ideas/wiki/Python-3.12-Goals)
   - no GIL: [Sam Gross pep for 3.13](https://github.com/colesbury/nogil) 
   - c-extensions: a PEP for Hpy c-interfacing
- Python ubiquity movement:
   - WASM initiatives (pyodide/pyscript.net/jupyterlite)
 - micromamba option

**Comment by stonebig** [comment]
audit trail:
- [ ] numba-0.56.0
- [ ] Pyside-6.3.2
- [ ] cpython-3.10.6

**Comment by firai** [comment]
>   * [ ]  BLAS move (if we loose cgohlke site)

While the UC Irvine LFD site hasn't been taken down yet, the site has already been archived and it seems like Gohlke has stopped compiling new wheels. For example, MKL wheels are not available for Numpy 1.23, which was released 3 days before the last update of the LFD site, or for Numpy 1.23.1.

**Comment by ivankravets** [comment]
Could you add binaries for Python 3.9.13?

**Comment by stonebig** [comment]
> Could you add binaries for Python 3.9.13?

There is no resources nor interest to maintain & update old cpython-3.x versions.

**Comment by RoyiAvital** [comment]
Has anyone tried to contact @cgohlke to know the status?
It will be a big lost, as the MKL is one of the best features of WinPython.

**Comment by firai** [comment]
He announced on that site a month or two ago that UC Irvine LFD lost funding, so the site would be shutting down before July.

**Comment by RoyiAvital** [comment]
I see. Too bad.

I hope hes shares his building scripts so it might be replicated.
Maybe even by a GitHub Actions or something?

**Comment by stonebig** [comment]
> Has anyone tried to contact @cgohlke to know the status? It will be a big lost, as the MKL is one of the best features of WinPython.

Hummm, looking at an Intel site, it seems that "pip install mkl" may offer a path forward.

**Comment by stonebig** [comment]
Couting MKL casualties ... at first look, it seems that:
- fetching MKL + numpy + scipy from pip shall work,
- but we loose "python-next" ability to get scientific wheels ahead of cPython-3.11 (or PyPy-3.9) official release of these wheels.

**Comment by stonebig** [comment]
b1 (2022-08-06/0x)

Changes from WinPython 2022-02:
- Infrastructure:
   - python-3.10.6
   - python-3.11.0rc1
- New Packages:
   - missingno, sspyrs
- Upgraded Packages: 
   - pyarrow-9.0, scikit-learn-1.1.2
   - SciPy-1.9.0 (on Python-3.10)
- variation per version:
   - for Python-3.10 : mkl is added and cgohkle wheels are progressively replaced per standard wheels, 
 
Areas of particular interest for testers:
- Latest Jupyterlab and Spyder improvements
- Numpy-1.22 has the data-api interface (more "standard" future)
 
Next build effort:
- can we tune again 7zip for size, (we grew 10% in size, not good), ... **hum, it's the compression inquiry experiments ... 7 compressed version of pandoc.exe, 56 Mo) included**, will re-spin
- the pyscript/pyodide eco-system, if it can generate a GUI,
- looking on PyPy-3.9 & PySide6-6.3.2 to be hoped mid june

**Comment by RoyiAvital** [comment]
@stonebig , Does it mean you were able to overcome the MKL issues?

**Comment by stonebig** [comment]
> @stonebig , Does it mean you were able to overcome the MKL issues?

Winpython64-3.10.6.1b1.exe of today is made with intel MKL standard wheel.
Not sure it's all perfect yet in every corner: testing and feedback is recommanded.

**Comment by RoyiAvital** [comment]
@stonebig , I will test it.
What I meant is that you wrote that the Intel distribution of NumPy / SciPy is behind the release cycle.
Did you find an alternative which is updated with each release of NumPy / SciPy and in a timely manner?

**Comment by stonebig** [comment]
> @stonebig , I will test it. What I meant is that you wrote that the Intel distribution of NumPy / SciPy is behind the release cycle. Did you find an alternative which is updated with each release of NumPy / SciPy and in a timely manner?

on "current" cpython, it shall remain timely. The problem is for PyPy or un-final-released cpython-3.11

**Comment by stonebig** [comment]
interlude (2022-08-0x): Python-3.11.0rc1

Changes from Python-3.11.0b4 :

- infrastructure:
   - Python-3.11.0rc1,
   - improved "Windows Terminal" icon launcher

**Comment by stonebig** [comment]
b2 (2022-08-1x)

Changes from WinPython 2022-03 b1:
- Infrastructure:
   - wppm rework to not use dying pkg_resources of pip, but  'pip inspect' instead
   - integrate dying pipdeptree features in wppm command line
         - wppm -p pandas = packages requirement for pandas  (at 2 levels of depth per default)
         - wppm -p pandas[test] = packages requirement for pandas[test]  (at 2 levels of depth per default)
         - wppm -l 3 -r pytest = packages upward dependancies to pytest at 3 levels upward 
- New Packages:
- Upgraded Packages: 
- variation per version:
   - for Python-3.10 : mkl is added as Christoph Gohkle wheels are progressively replaced per standard wheels,

**Comment by RoyiAvital** [comment]
@stonebig , When you say `MKL` is added, do you mean you add the wheel for MKL but also a version of `NumPy` and `SciPy` which are compiled with `MKL`, right?

**Comment by stonebig** [comment]
> @stonebig , When you say `MKL` is added, do you mean you add the wheel for MKL but also a version of `NumPy` and `SciPy` which are compiled with `MKL`, right?
> 
> Could you share the links to those wheels?

No,  the hope is that the standard  wheel of numpy see and choose the mkl. The mkl itself is a sll

**Comment by RoyiAvital** [comment]
@stonebig , I think it won't. I think one must build NumPy / SciPy in a way that utilize MKL.
Do you have any hint to think differently?

**Comment by stonebig** [comment]
Humm, maybe you're right,  https://www.intel.com/content/www/us/en/developer/articles/tool/oneapi-standalone-components.html#python

I don't know how to check mkl is used per numpy
If you can find à  heck test or command

**Comment by stonebig** [comment]
> @stonebig , When you say `MKL` is added, do you mean you add the wheel for MKL but also a version of `NumPy` and `SciPy` which are compiled with `MKL`, right?
> 
> Could you share the links to those wheels?

so the possible link is this one, https://pypi.org/project/intel-numpy/#files
... but it stops at python-3.9 ... you trade python-3.10 freshness and speed-up to get mkl.... wired

**Comment by RoyiAvital** [comment]
This is Intel NumPy, They recompile it with MKL. It seems your path is just add the MKL DLL's not a recompiled NumPy?

What I;m saying is that I'm not sure just adding the DLL is enough without recompilation. What do you think?

**Comment by stonebig** [comment]
> This is Intel NumPy, They recompile it with MKL. It seems your path is just add the MKL DLL's not a recompiled NumPy?
> 
> What I'm saying is that I'm not sure just adding the DLL is enough without recompilation. What do you think?

interesting thread here https://github.com/scipy/scipy/issues/11812

**Comment by RoyiAvital** [comment]
Yea, This is similar to LibTrampoline project at Julia. The question if this was implemented. It doesn't seem so.

**Comment by stonebig** [comment]
> Yea, This is similar to LibTrampoline project at Julia. The question if this was implemented. It doesn't seem so.

As I see numpy contributors in 'PyPy' and 'https://data-apis.org/array-api/latest/", I suppose:
- the problematic is being handled, 
- and  just keep upgrading to latest cpython and latest numpy/bumba compatible will get us to awesomeness.

These people are very busy making their Mac M1 & M2 working at maximum speed, so the 'MKL' thing can be percieved as 'passé'

**Comment by RoyiAvital** [comment]
Where did you get the notion it becomes a legacy? There is no alternative, yet at least, to MKL on the `x86` sphere where WinPython is targeting. OpenBLAS isn't close when you use LAPACK functions.
If WinPython switches to OpenBLAS it means the performance will deteriorate for numerical work much much more than the 30-40% achieved in Python 3.11 (Which won't help with numerical work).

What about replicating @cgohlke recipes to build NumPy and SciPy?
I tried approaching him, yet I did not get any response. Maybe you can try? He'll probably answer you.

**Comment by stonebig** [comment]
> Where did you get the notion it becomes a legacy?

'pip intel-numpy' is not available for python-3.10, from the company which shall be the most motivated to feed it to you.

**Comment by RoyiAvital** [comment]
It is because their distribution isn't popular.
Most of the people use Conda and they have NumPy with MKL (This is the default) on their repository.

So maybe using the package from Conda is the solution here?

**Comment by RoyiAvital** [comment]
Maybe in the long run, this might help: https://www.nsf.gov/awardsearch/showAward?AWD_ID=2209877.

**Comment by stonebig** [comment]
Will be done by july 2025, not exactly tomorrow.....

**Comment by RoyiAvital** [comment]
I wrote in the long run :-).
If we get the recipes from @cgohlke I will be trying to replicate them for WinPython. But can we get in touch with him? Have you tried?

**Comment by stonebig** [comment]
> I wrote in the long run :-). If we get the recipes from @cgohlke I will be trying to replicate them for WinPython. But can we get in touch with him? Have you tried?

No  There is no fun nor interest nor strategy  that would motivate such a move/burden. It can only be done and sustained  per a way bigger organization.

**Comment by RoyiAvital** [comment]
I don't need to recreate all the recipes he created, Only NumPy and SciPy.

**Comment by stonebig** [comment]
There is a new contender lfortran that smells interesting, to replace intel fortran:
- https://twitter.com/pyodide/status/1560948754303393793
- https://github.com/lfortran/lfortran

**Comment by RoyiAvital** [comment]
It is a compiler, It won't make OpenBLAS faster than MKL. In MKL the kernels are written in assembly. The magic isn't the Intel compiler, it is the kernels.

**Comment by stonebig** [comment]
> It is a compiler, It won't make OpenBLAS faster than MKL. In MKL the kernels are written in assembly. The magic isn't the Intel compiler, it is the kernels.

That is something only Christoph Gohlke, or Intel or Microsoft, can carry on with the required care, and forward-looking care in case of Christoph.

**Comment by RoyiAvital** [comment]
It is only a compilation recipe, nothing special here. We just need his recipe.

**Comment by RoyiAvital** [comment]
By the way, on Intel MKL NumPy: https://pypi.org/project/intel-numpy/ I can see how to install it from Conda using `pip`. It is updated to the latest. Could it be a solution?

**Comment by RoyiAvital** [comment]
What about out of the box idea?
Maybe build WinPython around [MicroMamba](https://github.com/mamba-org/mamba)?

It is a single file to start with, so all the scripts will be much easier around it.

**Comment by stonebig** [comment]
it's indeed a possible path

**Comment by RoyiAvital** [comment]
What do you mean by:

> I'm a bit unsettled per the server domain https://micro.mamba.pm

**Comment by stonebig** [comment]
pm = St Pierre & Miquelon

**Comment by stonebig** [comment]
the demise of https://www.lfd.uci.edu/~gohlke/pythonlibs/ and also indirectly the fall-of-the-tree of intel are problems.

**Comment by stonebig** [comment]
b3 (2022-09-18: Python updates)

Changes from WinPython 2022-03 b1:
- Infrastructure:
   - cPython-3.10.7,
   - cPython-3.11.0rc2 
 - New Packages:
   - duckdb: [fast] in-process "SQL" OLAP database management system (overs pandas, parquet, csv, ...)
- Upgraded Packages: 
   - msvc-runtime-14.32.31326
   - numpy-1.23.3 for python-3.10
   - numba-0.56.2 
   - pandas-1.4.4
- Removed Packages: 
- variation per version:
   - for Python-3.10 : mkl is added as Christoph Gohkle wheels are progressively replaced per standard wheels,

Areas of particular interest for testers:
- numba with numpy-1.23.3 on Python-3.10
- duckdb

Next build effort: rc
- Python-3.11.0 final
- some digging around duckdb and pyodide further progress

**Comment by hiccup7** [comment]
Do I understand correctly that WinPython-64bit v3.10.7.0 b3 uses pip to install numpy and scipy from the following links?
https://pypi.org/project/numpy/
https://pypi.org/project/scipy/

If so, why does WinPython-64bit v3.10.7.0 b3 contain scipy v1.9.0?  scipy v1.9.1 was released 2022-August-26.

**Comment by stonebig** [comment]
> Do I understand correctly that WinPython-64bit v3.10.7.0 b3 uses pip to install numpy and scipy from the following links? https://pypi.org/project/numpy/ https://pypi.org/project/scipy/
> 
> If so, why does WinPython-64bit v3.10.7.0 b3 contain scipy v1.9.0? scipy v1.9.1 was released 2022-August-26.

well, it's a miss

**Comment by stonebig** [comment]
b4 (2022-10-16: Python updates)

Changes from WinPython 2022-03 b3
- Infrastructure:
   - cPython-3.10.8,
   - a bug corrected in winpython pipdeptree replacement: "wppm -p" and "wppm -r"
   - upgraded some pyscript examples to 2022.09.1 syntax
 - New Packages:
   - xgboost (because in panel pyodide example)
- Upgraded Packages: 
   - numpy-1.23.4 for python-3.10
   - numba-0.56.3 , pandas-1.5.0
   - polars-0.14.19, panel_0.14.0, jupyterlab-3.4.8
- Removed Packages: 
- variation per version:
   - for Python-3.10 : mkl is added as Christoph Gohkle wheels are progressively replaced per standard wheels,

Areas of particular interest for testers:
- numba with numpy-1.23.4 on Python-3.10
- duckdb, polars ... : the 'next generation' of pandas,
- pyodide / pyscript examples

Next build effort: rc
- Python-3.11.0 final

**Comment by RoyiAvital** [comment]
I don't see why you add `MKL`.
Just adding it while using the regular `NumPy` and `SciPy` packages doesn't do a thing.
It just takes a volume but nothing is using it.

**Comment by stonebig** [comment]
> I don't see why you add `MKL`. Just adding it while using the regular `NumPy` and `SciPy` packages doesn't do a thing. It just takes a volume but nothing is using it.

I had a doubt. Did you experiment a bit with mamba ?

**Comment by RoyiAvital** [comment]
Yes. It worked like `conda` just much much faster.

It seems you'll be able to create a portable distribution around it much easier.

**Comment by stonebig** [comment]
xx  (2022-10-2x: )

Changes from WinPython 2022-03 b4
- Infrastructure:
   - fixed the Python-3.10.8 idle bug (by cpython)
 - New Packages:
   - waitress
- Upgraded Packages: 
   - anyio
- Removed Packages: 
   - mkl is removed: no native support per numpy,  
- variation per version:
   - 

Areas of particular interest for testers:

Next build effort: rc
- Python-3.11.0 final

**Comment by stonebig** [comment]
> Hello, it is possible to upgrade the python version of WinPython or do I have to reinstall it again? Sorry, if it not the right space, but I couldn't find any information about this topic.

Installing/unzipping a new version is the suggested method.

**Comment by stonebig** [comment]
xx  (2022-10-30: )

Changes from WinPython 2022-03 b5
- Infrastructure:
   - rework the register / unregister (experimental legacy Python way, not 'py' way):
     - register/unregister are wppm option, so don't need Qt
     - it still needs pywin32 to do more then propose to launch python when double-clicking on a .py file.
 - New Packages:
- Upgraded Packages: 
- Removed Packages: 
- variation per version:
   - "dot" version contains nothing but msvc_runtime, setuptools, wheel, winpython, and sqlite_bro

Areas of particular interest for testers:
- for the adventurous, "--register and "--unregister" options.

Next build effort: rc
- Python-3.11.0 final (and hopefully some missing packages will be build)

**Comment by stonebig** [comment]
post-rc remarks:
- IDLEX seems still working in Cpython-3.11.0, had issues during the summer, but don't remember where.
- dask-labextension suffers from a 'when-installed-per-pip' bug: https://github.com/dask/dask-labextension/issues/245

**Comment by RoyiAvital** [comment]
Maybe a good strategy would be waiting to Python 3.11.1. It will give time to the eco system to mature and have better compatibility.

**Comment by stonebig** [comment]
rc2 will just be with the dask-labextension update.

**Comment by stonebig** [comment]
> Maybe a good strategy would be waiting to Python 3.11.1. It will give time to the eco system to mature and have better compatibility.

In the particular case of Winpython:
- publishing is integral part of the maturation process,
- the ideal pace of WinPython publishing shall be every 2 month, following cPython update/maturation cycle,
- there should be at only one focus internal effort between release, to keep stability.
